### PR TITLE
PXB-2133 Wrong prtype stored for DECIMAL columns in .cfg file by export

### PR DESCRIPTION
**Pull request by rahulmalik87**
Problem:
Wrong prtype value is calculated by prepare during export

Analysis:
there are some  wrong fields calculated for different columns types
like collation_id, binary, column type, etc.

Fix:
calculate the fields the way server is calculating

### CODE REVIEW

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
Instead of maintaining this variant, check if the dd_table_open_dd_obj() variant can be used.

See if the stats funcationality is already utilizing this. If we can utiilize this it is awesome. Because that path, creates TABLE*, TABLE_SHARE and dd::Table to create dict_table_t.

This is the right way and bug-free way IMHO..

but it all depends if TABLE* and TABLE_SHARE* can be created.

Please debug and add your observations here

**Comment by rahulmalik87** [comment]
> Instead of maintaining this variant, check if the dd_table_open_dd_obj() variant can be used.
> 
> See if the stats funcationality is already utilizing this. If we can utiilize this it is awesome. Because that path, creates TABLE*, TABLE_SHARE and dd::Table to create dict_table_t.
> 
> This is the right way and bug-free way IMHO..
> 
> but it all depends if TABLE* and TABLE_SHARE* can be created.
> 
> Please debug and add your observations here

Even stats call dd_table_create_on_dd_obj not dd_table_open_on_dd_obj.
I think table_share object can only be invoked if can start Data dicitionary some way.

**Comment by satya-bodapati** [comment]
Can you please provide stack trace?

**Comment by rahulmalik87** [comment]
> Can you please provide stack trace?

#0  dd_table_create_on_dd_obj (dd_table=0x55555d3147e8, dd_part=0x0,
    schema_name=0x7fffffffd3c0, is_implicit=true)
    at /home/rahul.malik/MySQL/src/x8/storage/innobase/dict/dict0dd.cc:377
#1  0x0000555558d7d93f in dd_table_load_part (table_id=1058, dd_table=..., dd_part=0x0,
    table=@0x7fffffffd2e8: 0x0, thd=0x55555d30a970, schema_name=0x7fffffffd3c0, implicit=true)
    at /home/rahul.malik/MySQL/src/x8/storage/innobase/dict/dict0dd.cc:1071
#2  0x0000555558d7dd9b in dd_table_load_on_dd_obj (client=0x55555d30e1a0, space_id=1,
    dd_table=..., table=@0x7fffffffd2e8: 0x0, thd=0x55555d30a970, schema_name=0x7fffffffd3c0,
    implicit=true) at /home/rahul.malik/MySQL/src/x8/storage/innobase/dict/dict0dd.cc:1137
#3  0x000055555831ba10 in dict_load_tables_from_space_id (space_id=1, thd=0x55555d30a970,
    trx=0x7ffff53bf870)
    at /home/rahul.malik/MySQL/src/x8/storage/innobase/xtrabackup/src/xtrabackup.cc:2196
#4  0x000055555831be9c in dict_load_from_spaces_sdi ()
    at /home/rahul.malik/MySQL/src/x8/storage/innobase/xtrabackup/src/xtrabackup.cc:2253
#5  0x000055555831c0a2 in innodb_init (init_dd=true, for_apply_log=false)
    at /home/rahul.malik/MySQL/src/x8/storage/innobase/xtrabackup/src/xtrabackup.cc:2304
#6  0x00005555583215c8 in xtrabackup_stats_func (argc=9, argv=0x55555ce38278)
    at /home/rahul.malik/MySQL/src/x8/storage/innobase/xtrabackup/src/xtrabackup.cc:4446
#7  0x000055555832a7f7 in main (argc=11, argv=0x7fffffffe318)
    at /home/rahul.malik/MySQL/src/x8/storage/innobase/xtrabackup/src/xtrabackup.cc:7459
(gdb)

**Comment by satya-bodapati** [comment]
> > Can you please provide stack trace?
> 
> #0 dd_table_create_on_dd_obj (dd_table=0x55555d3147e8, dd_part=0x0,
> schema_name=0x7fffffffd3c0, is_implicit=true)
> at /home/rahul.malik/MySQL/src/x8/storage/innobase/dict/dict0dd.cc:377
> #1 0x0000555558d7d93f in dd_table_load_part (table_id=1058, dd_table=..., dd_part=0x0,
> table=@0x7fffffffd2e8: 0x0, thd=0x55555d30a970, schema_name=0x7fffffffd3c0, implicit=true)
> at /home/rahul.malik/MySQL/src/x8/storage/innobase/dict/dict0dd.cc:1071
> #2 0x0000555558d7dd9b in dd_table_load_on_dd_obj (client=0x55555d30e1a0, space_id=1,
> dd_table=..., table=@0x7fffffffd2e8: 0x0, thd=0x55555d30a970, schema_name=0x7fffffffd3c0,
> implicit=true) at /home/rahul.malik/MySQL/src/x8/storage/innobase/dict/dict0dd.cc:1137
> #3 0x000055555831ba10 in dict_load_tables_from_space_id (space_id=1, thd=0x55555d30a970,
> trx=0x7ffff53bf870)
> at /home/rahul.malik/MySQL/src/x8/storage/innobase/xtrabackup/src/xtrabackup.cc:2196
> #4 0x000055555831be9c in dict_load_from_spaces_sdi ()
> at /home/rahul.malik/MySQL/src/x8/storage/innobase/xtrabackup/src/xtrabackup.cc:2253
> #5 0x000055555831c0a2 in innodb_init (init_dd=true, for_apply_log=false)
> at /home/rahul.malik/MySQL/src/x8/storage/innobase/xtrabackup/src/xtrabackup.cc:2304
> #6 0x00005555583215c8 in xtrabackup_stats_func (argc=9, argv=0x55555ce38278)
> at /home/rahul.malik/MySQL/src/x8/storage/innobase/xtrabackup/src/xtrabackup.cc:4446
> #7 0x000055555832a7f7 in main (argc=11, argv=0x7fffffffe318)
> at /home/rahul.malik/MySQL/src/x8/storage/innobase/xtrabackup/src/xtrabackup.cc:7459
> (gdb)

This is startup path. Please comment this loading from sdi and then check how the stats path loads the tables?

File: storage/innobase/xtrabackup/test/t/xb_export.sh [hunk 1]
```diff
@@ -126,23 +131,35 @@ start_server $mysql_extra_args
...
+ls -tlr $backup_dir/incremental_sample/*cfg
 cfg_count=`find $backup_dir/incremental_sample -name '*.cfg' | wc -l`
-vlog "Verifying .cfg files in backup, expecting 5."
-test "$cfg_count" -eq 5
+vlog "Verifying .cfg files in backup, expecting 3. currently "$cfg_count
+if [ "$cfg_count" -ne 3  ]
+then
+    vlog "cfg files are not equal to 3 "
+    exit -1
+fi
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/test/t/xb_export.sh]: why the count decreased from 5 to 3?

File: storage/innobase/dict/dict0dd.cc [hunk 2]
```diff
@@ -635,6 +637,18 @@ dict_table_t *dd_table_create_on_dd_obj(const dd::Table *dd_table,
...
+      case MYSQL_TYPE_TINY_BLOB:
+        field_type = MYSQL_TYPE_BLOB;
+        break;
+      case MYSQL_TYPE_NEWDATE:
+        field_type = MYSQL_TYPE_DATE;
+        break;
+      case MYSQL_TYPE_ENUM:
+      case MYSQL_TYPE_SET:
+        field_type = MYSQL_TYPE_STRING;
+        break;
```
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dd.cc:675]: where is the conversion done for  running server??

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/dict/dict0dd.cc (hunk 2 above)
Comment (rahulmalik87) [reply storage/innobase/dict/dict0dd.cc:675]: sql/field.h
 enum_field_types type() const final override { return MYSQL_TYPE_STRING; }

File: storage/innobase/xtrabackup/test/t/xb_export.sh (hunk 1 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/t/xb_export.sh]: fixed the test case. some confusion because of test.

**Review by satya-bodapati — APPROVED** [review]

**Pull request merged**